File: src/dataset.cpp
#include <pybind11/pybind11.h>

#include <cassert>
#include <fstream>
#include <iostream>
#include <sstream>

#include "dataset.hpp"
#include "subject.hpp"
#include "utils.hpp"

Dataset::Dataset(const std::filesystem::path& bids_dir, bool silent)
    : bids_dir(bids_dir), silent_(silent) {
  // Load participants sidecar
  std::ifstream sidecar_file(this->participants_sidecar_filepath());
  if (sidecar_file.is_open()) {
    sidecar_file >> this->participants_sidecar_;
    sidecar_file.close();
  } else {
    std::cerr << "Could not open participants sidecar file." << std::endl;
  }

  // Load participant table
  this->load_participants_table_();
}

std::optional<Subject> Dataset::add_subject(
    const std::unordered_map<std::string, std::string>& args) {
  if (args.size() > this->participants_properties().size()) {
    std::cerr << "Too many arguments" << std::endl;
    throw std::runtime_error("AssertionError: Too Many Arguments");
    return std::nullopt;
  }

  if (!this->silent_ &&
      !this->confirm_add_subject_(std::stoi(args.at("participant_id")),
                                  args.at("name"))) {
    return this->get_subject(std::stoi(args.at("participant_id")));
  }

  Subject subject(this->shared_from_this(), args);
  if (std::find(this->get_subjects().begin(), this->get_subjects().end(),
                std::stoi(subject.get_participant_label())) ==
      this->get_subjects().end()) {
    this->append_participant(subject);
  }

  std::filesystem::create_directories(this->bids_dir /
                                      subject.get_participant_id());
  this->load_participants_table_();
  return subject;
}

void Dataset::append_participant(const Subject& subject) {
  std::ofstream participant_file(participants_filepath(), std::ios::app);
  if (participant_file.is_open()) {
    for (const auto& [key, value] : subject.to_dict()) {
      participant_file << key << "\t" << value << "\n";
    }
    participant_file.close();
  } else {
    std::cerr << "Could not open participants file for writing." << std::endl;
  }
}

std::optional<Subject> Dataset::get_subject(int idx) const {
  auto participant_id = Subject::ensure_participant_id(idx);
  auto it = std::find_if(
      this->participants_table_.begin(), this->participants_table_.end(),
      [&participant_id](
          const std::unordered_map<std::string, std::string>& participant) {
        for (auto& [k, v] : participant)
          return participant.at("participant_id") == participant_id;
      });

  if (it != participants_table_.end()) {
    return Subject(this->shared_from_this(), *it);
  } else {
    return std::nullopt;
  }
}

std::vector<int> Dataset::get_subjects() const {
  std::vector<int> subjects;
  for (const auto& participant : participants_table_) {
    subjects.push_back(std::stoi(participant.at("participant_id").substr(4)));
  }
  return subjects;
}

bool Dataset::is_subject(int idx) const { return get_subject(idx).has_value(); }

bool Dataset::confirm_add_subject_(int subject_idx,
                                   const std::string& subject_name) {
  if (!this->is_subject(subject_idx)) {
    std::string prompt = std::string(
        "New Participant ID\nPlease Confirm that this is a new participant.");
    py::object qprompt = py::module_::import("libbids").attr("qprompt");
    return qprompt(prompt, "OK", "No. No this participant is not new")
        .cast<bool>();
  } else {
    auto subject = get_subject(subject_idx);
    if (subject) {
      assert(subject_name == subject->get_participant_name());
    }
    return false;
  }
}

// Properties
std::vector<std::string> Dataset::participants_properties() const {
  std::vector<std::string> properties;
  properties.push_back("participant_id");
  for (const auto& member : this->participants_sidecar_.getMemberNames()) {
    properties.push_back(member);
  }
  return properties;
}

std::filesystem::path Dataset::participants_filepath() const {
  return bids_dir / "participants.tsv";
}

std::unordered_map<std::string, std::string> Dataset::participants_sidecar()
    const {
  std::unordered_map<std::string, std::string> sidecar_map;
  for (const auto& member : participants_sidecar_.getMemberNames()) {
    sidecar_map[member] = participants_sidecar_[member].asString();
  }
  return sidecar_map;
}

std::filesystem::path Dataset::participants_sidecar_filepath() const {
  return bids_dir / "participants.json";
}

std::vector<std::unordered_map<std::string, std::string>>
Dataset::participant_table() const {
  return this->participants_table_;
}

void Dataset::load_participants_table_(void) {
  if (!std::filesystem::exists(this->participants_filepath())) return;

  std::ifstream participants_fs(this->participants_filepath());
  std::string header_line(255, '\0');
  participants_fs.getline(header_line.data(), header_line.size());
  header_line = trim(header_line);
  std::istringstream header_stream(header_line);
  std::vector<std::string> header;
  while (!header_stream.eof()) {
    std::string word(255, '\0');
    header_stream.getline(word.data(), word.size(), '\t');
    word = trim(word);
    bool found = false;
    for (auto const& i : this->participants_properties()) found |= (i == word);
    assert(found);
    header.push_back(word);
  }

  while (!participants_fs.eof()) {
    std::string line(255, '\0');
    participants_fs.getline(line.data(), line.size());
    line = trim(line);

    std::istringstream line_stream(line);
    std::unordered_map<std::string, std::string> row;
    int i = 0;
    while (!line_stream.eof()) {
      std::string value(255, '\0');
      line_stream.getline(value.data(), value.size(), '\t');
      value = trim(value);
      if (!value.empty()) {
        std::string& key = header[i];
        row.emplace(std::pair(key, value));
        i++;
      }
    }
    if (row.size() > 1) this->participants_table_.push_back(row);
  }
}

File: libbids/dataset.py
"""
Supplemental pybids object for manipulating the dataset
"""
import os
import sys
import json
import pandas as pd  # type: ignore
from bids.layout import BIDSLayout  # type: ignore
from pathlib import Path
from typing import Dict, List, Optional, cast
from .clibbids import Subject

try:
    from PyQt6.QtWidgets import QMessageBox, QWidget  # type: ignore
except ModuleNotFoundError:
    from PyQt5.QtWidgets import QMessageBox, QWidget  # type: ignore


class Dataset:
    def __init__(self, bids_dir: Path, silent: bool = False):
        """Initialize a BIDSDataset

        Parameters
        ----------
        bids_dir : Path
            Path to the root directory of the BIDS layout
        """
        self.bids_dir: Path = bids_dir
        self.layout: BIDSLayout = BIDSLayout(str(self.bids_dir), validate=False)
        self.silent: bool = silent

    def add_subject(self, *args, **kwargs) -> Optional[Subject]:
        """add a subject to the dataset. the arguments passed to this function
        are defined by the JSON sidecar of participants

        Returns
        -------
        Optional[Subject]
            If the subject is successfully added, the subject object is
            returned, else None
        """
        assert len(args) + len(kwargs) <= len(
            self.participants_properties
        ), "Too many arguments"
        if len(args) > 0:
            keys: List[str] = self.participants_properties[: len(args)]
            kwargs.update({k: v for k, v in zip(keys, args)})

        if (not self.silent) and (
            not self._confirm_add_subject(kwargs["participant_id"], kwargs["name"])
        ):
            return self.get_subject(kwargs["participant_id"])

        subject = Subject(self, kwargs)
        if subject.get_participant_label() not in self.layout.get_subjects():
            self.append_participant(subject)

        os.makedirs(
            os.path.join(self.layout.root, subject.get_participant_id()), exist_ok=True
        )
        self.layout = BIDSLayout(self.layout.root, validate=False)
        return subject

    def append_participant(self, subject: Subject):
        """Add the given subject to our participants table

        Parameters
        ----------
        subject : Subject
            The subject object that has our subject's data
        """
        fn: str = "participants.tsv"
        fp: str = os.path.join(self.layout.root, fn)
        subject_data: dict = subject.to_dict()
        subject_table: pd.DataFrame = pd.DataFrame(pd.Series(subject_data))
        participant_data: pd.DataFrame = pd.concat(
            [self.participant_table, subject_table], ignore_index=True
        )
        participant_data.to_csv(fp, sep="\t", index=False)

    def get_subject(self, idx: int) -> Optional[Subject]:
        """Given a subject's index, get that subject or None if not found

        Parameters
        ----------
        idx : int
            The id of the subject

        Returns
        -------
        Optional[Subject]
        """
        participant_id: str = Subject.ensure_participant_id(idx)
        query = self.participant_table.query(f"participant_id == '{participant_id}'")
        if query.shape[0] < 1:
            return None

        data_dict: Dict = {k: str(v) for k, v in query.iloc[0, :].to_dict().items()}
        return Subject(self, data_dict)

    def get_subjects(self) -> List[int]:
        """Retreive a list of subject ids

        Returns
        -------
        List[int]
        """
        participant_ids: List[str] = self.participant_table["participant_id"].values.tolist()
        return [
            int(ids[4:])
            for ids in participant_ids
        ]

    def is_subject(self, idx: int) -> bool:
        """determins if the subject with the provided id exists

        Parameters
        ----------
        idx : int
            subject integer

        Returns
        -------
        bool
            Whether the subject exists
        """
        return self.get_subject(idx) is not None

    def refresh(self):
        """The lawyout needs to be refereshed when changes are made"""
        self.layout = BIDSLayout(self.layout.root, validate=False)

    def _confirm_add_subject(self, subject_idx: int, subject_name: str) -> bool:
        """Confirm that the subject should be added

        Parameters
        ----------
        subject_idx : int
            The id of the subjec to add
        subject_name : str
            The name of the subject. If this is not a new subject, we double
            check the name to ensure there is no overwriting

        Returns
        -------
        bool
            Whether to add the subject or not
        """
        if not self.is_subject(subject_idx):
            mb = QMessageBox.warning(
                cast(QWidget, None),
                "Warning",
                "New Participant ID\nPlease confirm that this is a new participant",
                QMessageBox.StandardButton.Ok,
                QMessageBox.StandardButton.Cancel,
            )
            if mb == QMessageBox.StandardButton.Cancel:
                return False
            else:
                return True
        else:
            subject: Subject = cast(Subject, self.get_subject(subject_idx))
            assert subject_name == subject["name"], (
                f"The participant's name ({subject_name}) "
                f"doesn't match the id ({subject['name']})"
            )
            return False

    @property
    def participants_properties(self) -> List[str]:
        return ["participant_id"] + list(self.participants_sidecar.keys())

    @property
    def participants_filepath(self) -> Path:
        return self.bids_dir.joinpath("participants.tsv")

    @property
    def participants_sidecar(self) -> Dict:
        return json.load(open(self.participants_sidecar_filepath, "r"))

    @property
    def participants_sidecar_filepath(self) -> Path:
        return self.bids_dir.joinpath("participants.json")

    @property
    def participant_table(self) -> pd.DataFrame:
        participant_query: List = self.layout.get(
            suffix="participants", extension="tsv"
        )
        if len(participant_query) < 1:
            return pd.DataFrame(columns=self.participants_properties)

        return pd.read_csv(self.participants_filepath, sep="\t")
